frontend/src/test/scala/bloop/dap/DebugProtocolSpec.scala

```diff
@@ -0,0 +1,197 @@
+           |    Thread.sleep(10000)
+           |  }
+           |}
+           |""".stripMargin
+
+      val logger = new RecordingLogger(ansiCodesSupported = false)
+      val project = TestProject(workspace, "p", List(main))
+
+      loadBspState(workspace, List(project), logger) { state =>
+        val output = state.withDebugSession(project, mainClassParams("Main")) { client =>
+          for {
+            _ <- client.initialize()
+            _ <- client.launch()
+            _ <- client.configurationDone()
+
+            previousSession <- client.restart()
+
+            _ <- client.launch()
+            _ <- client.configurationDone()
+            _ <- client.disconnect()
+
+            previousSessionOutput <- previousSession.allOutput
+          } yield previousSessionOutput
+        }
+
+        assertNoDiff(output, "")
+      }
+    }
+  }
+
+  test("picks up source changes across sessions") {
+    val correctMain =
+      """
+        |object Main {
+        |  def main(args: Array[String]): Unit = {
+        |    println("Non-blocking Hello!")
+        |  }
+        |}
+     """.stripMargin
+
+    TestUtil.withinWorkspace { workspace =>
+      val main =
+        """|/main/scala/Main.scala
+           |object Main {
+           |  def main(args: Array[String]): Unit = {
+           |    println("Blocking Hello!")
+           |    synchronized(wait())
+           |  }
+           |}
+           |""".stripMargin
+
+      val logger = new RecordingLogger(ansiCodesSupported = false)
+      val project = TestProject(workspace, "p", List(main))
+
+      loadBspState(workspace, List(project), logger) { state =>
+        // start debug session and the immediately disconnect from it
+        val blockingSessionOutput = state.withDebugSession(project, mainClassParams("Main")) {
+          client =>
+            for {
+              _ <- client.initialize()
+              _ <- client.launch()
+              _ <- client.configurationDone()
+              output <- client.firstOutput
+              _ <- client.disconnect()
+            } yield output
+        }
+
+        assertNoDiff(blockingSessionOutput, "Blocking Hello!")
+
+        // fix the main class
+        val sources = state.toTestState.getProjectFor(project).sources
+        val mainFile = sources.head.resolve("Main.scala")
+        Files.write(mainFile.underlying, correctMain.getBytes(StandardCharsets.UTF_8))
+
+        // start the next debug session
+        val output = state.withDebugSession(project, mainClassParams("Main")) { client =>
+          for {
+            _ <- client.initialize()
+            _ <- client.launch()
+            _ <- client.configurationDone()
+            _ <- client.exited
+            _ <- client.terminated
+            _ <- client.disconnect()
+            output <- client.allOutput
+          } yield output
+        }
+
+        assertNoDiff(output, "Non-blocking Hello!")
+      }
+    }
+  }
+
```

Comment:
Can we add a test when we restart the execution of a test suite while the program is still running tests suites? Maybe also think of other test cases we might want to be testing? Like what happens when things go wrong?

Reply:
`DebugServerSpec.restarting closes current client and debuggee` should verify that both the connection and the debuggee (a task obtained from `TestTask.runTestSuites`) are cancelled.
It's true, I have not verified if this task cancels correctly - I will look into it soon.

Reply:
in short, I was thinking of `DebugServerSpec` as a test suite which abstracts over the debuggee and verifies the server itself - something closer to unit tests than integration tests. `DebugProtocolSpec` is used to verify that the actual protocol works correctly - those are the "real" integration tests.